### PR TITLE
Add draft github pull-request template

### PR DESCRIPTION
**Pull request by leifdenby**
The aim of this PR is to add a pull-request template for use on github when add pull-requests to the common repo. The aim of my initial commit is to start the discussion on how we would like pull-requests to progress when collaborating on neural-lam.

As you will see from the template I have created so far the aim is to:

- create placeholders for the PR creator to add context so we know what the PR is for
- detail the steps of getting the PR reviewed and merged into the repo

Looking forward to hearing your thoughts all :)

Here's a direct link so you can see it rendered: https://github.com/leifdenby/neural-lam/blob/maint/add-github-pr-template/.github/pull_request_template.md

### CODE REVIEW

File: .github/pull_request_template.md
Comment (sadamov) [inline .github/pull_request_template.md:12]:
```suggestion
- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
```

**Review by sadamov — COMMENTED** [review]
Thanks, this will prove very useful! I added some smaller discussion points.

File: .github/pull_request_template.md
Comment (SimonKamuk) [reply .github/pull_request_template.md:18]: I completely agree with the added README suggestion, but I am not entirely convinced we should discourage the use of inline comments. Maybe the text could be changed to something like 
"I have commented my code where necessary, for example in hard-to-understand areas"

**Comment by SimonKamuk** [comment]
Great idea with the template, and good suggestions by @sadamov. I have just added a small comment above, and have no further changes - great work!

File: .github/pull_request_template.md
Comment (sadamov) [reply .github/pull_request_template.md:7]: This is actually not required, as it can be specified directly in the side-menu

File: .github/pull_request_template.md
Comment (joeloskarsson) [inline .github/pull_request_template.md:21]:
```suggestion
- [ ] I have given the PR a name that clearly describes the changes, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form).
```

My idea here is that we should be able to use the PR title directly as the commit message when merging (so the assignee does not have to spend extra time formulating a descriptive commit message when merging). I think it's nice to follow the convention of imperative mood for commit messages.

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:40]: I view this similarly as the checklist for reviewers as not really needed in this template.

**Review by joeloskarsson — CHANGES_REQUESTED** [review]

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:13]: Would actually err on the side not needing this. If a file-by-file description is needed then I would suggest the PR is trying to do too much

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:7]: Ok, we can put this in. We can always revise this template later if we don't use parts of it

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:18]: How about we write:

```markdown
- [ ] I have added docstrings to any new/modified functions
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
```

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:28]: we can also leave "easy to maintain" out? The other three cover this some degree and this isn't explicit enough about how we measure this that the comment is probably superfluous

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:12]: I love both additions, thank you!

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:14]: I'd be fine with leaving it out for that reason

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:23]: Yes, these steps spell out the review process, but I do think it is useful to see how far along the review process a given PR is. And also, as you say, reviewing guidelines should go somewhere. Putting the in the PR ensures that people have them when they need them too, and are more likely to use them too. But maybe this points to the need to have a brief discussion about the review process and then we can decide where to put it? Having them in the PR template has worked well for me before, but I am also happy to discussing doing this differently :)

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:19]: I think that is good. Test (linting) should catch missing docstrings right? So that is maybe superfluous if test pass?

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:35]: > I guess a drawback with just editing the changelog immediately is that what should be written there could change after discussions in the PR, but I think that is quite rare, and then you just have to remember to update the changelog edit as well.

Yes, to me this is the key drawback. Sometimes people will want to fix/change/add something and they way they have achieved that might change as the review process progresses. In my experience this is often the case. Also, people often forget to update the changelog so this is a a little last-minute reminder to do that before merging. And finally, a given PR doesn't have a URL until it has been created, and so updating the changelog with a separate commit after the PR as been created is typically necessary.

Comment (leifdenby) [reply .github/pull_request_template.md:35]: > A question about the changelog sections: This here seems to apply that one should add arbitrary new section if there is not a new one? But below the changelog sections are described as fixed (_added_, _changed_, _fixed_). What is the general idea for this? If we are sticking to explicitly the sections listed at https://keepachangelog.com/en/1.1.0/, then I think we should make that clear also here. One way to do that is to always have those sections under _unreleased_, and then there is never a need to add sections with a PR. Empty sections at the time of a release can just be removed.

Yes, I think having fixed sections sounds like a good idea. I was going to suggest we discuss what these should be so that we are on the same page about it. One section that might be good to include (in addition to *added, changed, fixed*) is *maintenance* (these are things that don't change fhe functionality of the code at all, but are say necessary package updates because something breaks, fixing the ci/cd linting, testing etc update). We can also put those in the other sections since they will often *fix* something, but sometimes this info isn't relevant to you if you just want to decide whether to update or not.

**Comment by leifdenby** [comment]
I am going to promote this to an actual PR from a draft PR now the discussion is well underway :smile:

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md]: The documentation entry was not added correctly now.

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:35]: I'm ok with keeping it like this then. Just make the changelog categories explicit and consistent in the template, _added, changed, fixed, maintenance_ sound good to me.

File: .github/pull_request_template.md
Comment (joeloskarsson) [inline .github/pull_request_template.md:22]: There's a missing )
```suggestion
- [ ] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
```

**Comment by joeloskarsson** [comment]
Any updates on this @leifdenby? Are there specific points you want more input on? Out of my points it is mainly the checkbox for adding tests that requires some further discussion. It would be good to try to merge this soon.

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:19]: > Test (linting) should catch missing docstrings right? So that is maybe superfluous if tests pass?

True, the linting checks that docstrings have been written, but not that they are useful. Maybe we should leave this in to just remind people to think through the docstrings?

File: .github/pull_request_template.md
Comment (leifdenby) [reply .github/pull_request_template.md:29]: I will remove "easy to maintain" since that is hard to measure objectively, we can always adapt this template later if we find it is missing something or doesn't quite work :)

**Comment by leifdenby** [comment]
TODO

- [x] remove line about having added tests
- [x] add option for documentation under the type of change
- [x] update changelog sections to choose from to be explicitly listed as *added, changed, fixed, maintenance*

**Comment by leifdenby** [comment]
Ok, I think this is ready to go now @joeloskarsson. Do you mind giving it a last read over?

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:24]: Yes, I think that's good. Proposal:
```suggestion
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
```

File: .github/pull_request_template.md
Comment (joeloskarsson) [inline .github/pull_request_template.md:35]:
```suggestion
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain
```

File: .github/pull_request_template.md
Comment (joeloskarsson) [inline .github/pull_request_template.md:48]:
```suggestion
- [ ] the tests pass
```

File: .github/pull_request_template.md
Comment (joeloskarsson) [reply .github/pull_request_template.md:24]: Update the text about doc-strings

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Gave it a proofread and fixed some grammar. After these fixes + updating the docstring text this should be good to merge!

**Comment by leifdenby** [comment]
Thanks for the thorough review @joeloskarsson!

**Comment by leifdenby** [comment]
This just needs your approval and then I will merge in :)

**Review by joeloskarsson — APPROVED** [review]
Great, all looking good now!

**Pull request merged**
